### ISSUE TITLE
<podcast:authorization>

### ISSUE THREAD

**Issue by tomrossi7** [Podcastindex-org/podcast-namespace#518]
I propose we create a tag for podcast hosts to include an authorization_uri for authorizing ownership of a podcast.

Something like this:
```
<podcast:authorization>
  https://www.mypodcasthost.com/authorize/my/parameters
</podcast:authorization>
```

Now anyone that ingests the RSS feed and wants to verify ownership, can pop a web browser to the authorization_uri. The request must include a redirect_uri (e.g. https://www.mypodcasthost.com/authorize/my/parameters?redirect_uri=https://myplayer.com/authorization). The host will receive the request and walk the podcaster through authorization. Once its completed, they redirect to the redirect_uri.

I'm not a mobile developer, but I believe this would work with [Universal Links](https://medium.com/bumble-tech/universal-links-for-android-and-ios-1ddb1e70cab0) as the redirect_uri.

This is basically a dumbed down version of OAUTH, but would be simple and powerful.

**Comment by theDanielJLewis** [comment]
Why not use `<podcast:txt>` for this?

**Comment by jamescridland** [comment]
> Why not use `<podcast:txt>` for this?

If I understand this correctly, it *does* use `podcast:txt`. Here's the user flow:

## Currently, without `podcast:authorization`

"Thanks for wanting to claim this podcast on MyNewService. In order to claim it, please visit your podcast hosting company, discover where it lets you change the `podcast:txt` value, and add the obscure code `C8ED` to the `podcast:txt` value, which might be comma-delimited with other elements. The `podcast:txt` feature may be called Validation, or something else. Who knows. Then, come back to this web page, and hopefully by then the RSS feed will have been regenerated although who knows. We'll then check it and your podcast will be claimed."

## With `podcast:authorization`

"Thanks for wanting to claim this podcast on MyNewService. `CLICK THIS LINK` to claim it with your podcast host."

Have I got this right, @tomrossi7 ?

**Comment by tomrossi7** [comment]
> Why not use <podcast:txt> for this?

The `txt` tag is inspired by the DNS txt tag which is an absolute generic swiss-army knife tag to capture anything that may not be captured already in other standards (eg SPF, DMARC, authorization). As @jamescridland pointed out, its a perfect tag to use to just include some random code that someone for one reason or another needs to see in your RSS feed. The `authorization` tag is very specifc. This tag does one thing: provide a URL for authorization.

@jamescridland yes! That's exactly the kind of flow you could create with this tag. This completely eliminates the need for the podcaster to add anything to their RSS feed and immediately provide authorization to any app seeking to verify ownership.

**Comment by mkadin** [comment]
### Some Security Stuff
After the host auths you, and you are redirected back, the site that receives you needs to do some kind of handshake with the host with a token from the redirect URL, right?  Otherwise, this is easy to fake?

Also, I think to avoid phishing or some other security problems, the host would have to have an allowlist of sites that the redirect can go to.  Here's the attacking scenario:

a) Bad guy wants to steal your podcast on listening app X that supports <podcast:authorization>
b) Good guy gets a phishing email "Sign up to make mad cash on listening app X"
c) Good guy clicks the link, sees the auth page for his host, at its authentic URL, completes the login.
d) Redirect goes to bad guy's domain, steals confirmation token, which they then use to take over ownership.

Google, Stripe, etc, all have allow lists for domains and/or sites that the redirect can go to.  I think hosts would have to do the same.  This starts to increase the complexity of this then, where each host needs a list of approved apps.

So I think the actual implementation of this is a little bit harder / more complicated than your description.  It is, of course, doable!

### Separately, my personal view
The problem that @jamescridland describes is more of a branding problem / the difference between a feature's name and how it's rendered in the feed.  I feel 95% of the problem with <podcast:txt> would be standardizing the podcaster-facing name and the language we use around it.

The average of times a podcaster connects their podcast to a new service is maybe < 20 over the lifetime of their podcast, probably much less.  The pattern of 'enter a code' is a very common one for verification around the internet, and is simple for hosts and apps to implement.  Apple, one of if not the largest apps, already uses this technique.  I'd love to see more iteration on the existing solution vs. tossing it just yet.  Just my 2 cents.

**Comment by tomrossi7** [comment]
> the site that receives you needs to do some kind of handshake with the host with a token from the redirect URL, right?

No I don't think because this is a one-time authorization. The only reason I can think of for a token would be if we wanted them to continue to make authorized requests.

>  I feel 95% of the problem with podcast:txt would be standardizing the podcaster-facing name and the language we use around it.

I think 99% of the problem is complexity for the podcaster (and the podcast host) that significantly interrupts the flow of the podcast app that requires authorization.

**Comment by samsethi** [comment]
As a podcast host.  I fully support the removal of the email as it was used by bad actors aka Acast to spam hosts.  Podcast:Txt  seemed to be a simple solution at first glance but in practical terms it is a nightmare because it relies on the user to login to their host find the Podcast:txt field and then republish the RSS feed.  This can take 5+ mins or more by which time they are bored or have gone onto another task. Tom's solution is simple and neat. We will try and test this with Buzzcast to see how it works.

**Comment by CharlesWiltgen** [comment]
> I fully support the removal of the email as it was used by bad actors aka Acast to spam hosts.

This proposal doesn't address that, since Apple requires a correct email for `<itunes:owner>`.

**Comment by tomrossi7** [comment]
> Apple (for example) requires a correct email for `<itunes:owner>`.

Apple no longer has this requirement.

**Comment by mkadin** [comment]
>I think 99% of the problem is complexity for the podcaster (and the podcast host) that significantly interrupts the flow of the podcast app that requires authorization.

I agree that an oauth-style flow would be a lot easier for the podcaster, for sure.  I'm just not sure the juice is worth the squeeze -- would love to hear from more folks.  Our customer bases likely have varying characteristics / struggles.

>No I don't think because this is a one-time authorization. The only reason I can think of for a token would be if we wanted them to continue to make authorized requests.

I could be wrong or missing something, but imagine this scenario:

* Attacker wants to take your podcast on myplayer.com.  Starts the oauth-like process.  Gets a URL which looks like `https://www.mypodcasthost.com/authorize/my/parameters?redirect_uri=https://myplayer.com/authorization/some-custom-token`
* Attacker doesn't have username / pass at mypodcasthost.com, but can just copy and paste `https://myplayer.com/authorization/some-custom-token` into their browser and they're good to go.  The referer header can be easily spoofed.

I think for this transaction to be secure, there'd need to be some back end exchange between myplayer.com and mypodcasthost.com after the redirect, or some kind of signature in the redirect URL using a key that myplayer can authenticate.

**Comment by CharlesWiltgen** [comment]
> > Apple (for example) requires a correct email for `<itunes:owner>`.
> 
> Apple no longer has this requirement.

@tomrossi7, can you point me to a more current reference than what appears to be their canonical requirements? https://help.apple.com/itc/podcasts_connect/#/itcb54353390

**Comment by tomrossi7** [comment]
@mkadin I think you are correct 😬. A nefarious actor could just monitor the requests and they would see the `redirect_uri` being passed. Then they could just copy and paste it into a browser and spoof the authorization.

Okay, okay, okay. Here me out:
The player could direct the browser to the authorization URL and pass in a token. Then the person authorizing would have to login. Once they login, the host now knows the token has been authorized. The host could:
1. Add the token to the RSS feed `<podcast:txt>` or
2. Respond to HTTP GET requests with the current authorized tokens. This is similar to what I saw @CharlesWiltgen just posted!

Idea 1 would work today and be pretty easy BUT I hate that you are waiting for the RSS feed to update.
Idea 2 would require an additional endpoint on the podcast host.

Thoughts?

@CharlesWiltgen I don't know that Apple has formally communicated it anywhere, but many hosts (including Buzzsprout) have removed emails without any issues with Apple.

**Comment by samsethi** [comment]
Is there any further thoughts on this proposal.  I really want to implement this but don't want to waste my time if the proposal from (@tomrossi7) Tom Rossi is going to be changed.
